ffi/src/repr_c.rs

```diff
@@ -1,31 +1,36 @@
-#![allow(trivial_casts, clippy::undocumented_unsafe_blocks)]
-
-//! Logic related to the conversion of IR types to equivalent robust C types. Most of the
-//! types that implement [`Ir`], i.e. provide conversion into IR, will get an automatic
-//! implementation of [`CType`] and consequently an implementation of [`FfiType`]
+#![allow(trivial_casts)]
 
+//! Logic related to the conversion of IR types to equivalent robust C types. Types that are mapped into
+//! one of the predefined [`Ir`] types, will be provided an automatic implementation of traits in this module.
+//!
+//! Traits in this module mainly exists to bridge the gap between IR and C type equivalents. User should
+//! only implement these traits if none of the predefined IR types provide an adequate mapping
 use alloc::{boxed::Box, vec::Vec};
-use core::{hint::unreachable_unchecked, mem::ManuallyDrop};
+use core::{alloc::Layout, hint::unreachable_unchecked, mem::ManuallyDrop, ptr::addr_of_mut};
 
 use crate::{
     ir::{Ir, Opaque, Robust, Transmute, Transparent},
     slice::{OutBoxedSlice, SliceMut, SliceRef},
-    FfiConvert, FfiOutPtr, FfiReturn, FfiType, OutPtrOf, ReprC, Result,
+    FfiConvert, FfiOutPtr, FfiOutPtrRead, FfiOutPtrWrite, FfiOutput, FfiReturn, FfiType, LocalRef,
+    LocalSlice, ReprC, Result,
 };
 
-/// A type that can be converted into some C type. When implemented on [`Ir::Type`] the blanket
-/// implementations ensures that [`FfiType`] is also implemented on the type implementing [`Ir`].
-/// This trait mainly exists to move types of internal representation into C type equivalents.
+crate::decl_ffi_fn! { dealloc }
+
+/// A type that can be converted into some C type.
+///
+/// If a type also implements [`Ir`], i.e. has a defined internal representation, a blanket
+/// implementation of [`FfiType`] will be provided.
 pub trait CType<S> {
     /// C type current type can be converted into
     type ReprC: ReprC;
 }
 
-/// Conversion utility [`CType`]s. When implemented on [`Ir::Type`] the blanket implementation
-/// ensures that [`FfiType`] is also implemented on the type implementing [`Ir`]. This trait
-/// mainly exists to move types of internal representation into C type equivalents. When
-/// implementing make sure that the implementation doesn't conflict with [`NonLocal`]
-pub trait CTypeConvert<'itm, S, C>: Sized {
+/// Facilitates conversion of rust types to/from `ReprC` types.
```

Comment:
```suggestion
/// The trait facilitates conversion of rust types to/from `ReprC` types.
```

similar in other descriptions that start with "facilitates"